refactor error/warning reporting

adriaanm opened the pull request:
Refactor error/warning reporting to prepare making it configurable.

review by @gkossakowski (on hangout?)

adriaanm committed:
Pure refactor: reorder definitions

adriaanm committed:
Track symbol that caused a deprecation warning.

So that we can filter deprecations based on defining package.

Configurable error reporting will support a rule like:
"In compilation unit X, escalate deprecation warnings that
result from accessing members in package P that have been deprecated
since version V. Report an error instead of a warning for those."

TODO: remove deprecationWarning overload that doesn't take a `Symbol`?
(Replace by a default value of `NoSymbol` for the deprecated symbol arg?)

adriaanm committed:
Unclutter Reporter. Move truncation to ReplReporter.

Refactor to reduce the Reporter interface. Working towards
minimal interfaces in scala.reflect.internal that can be consumed by sbt/IDE/....
The scala.tools.nsc package is entirely private to the compiler (in principle).

A `Reporter` should only be used to inform (info/warning/error). No state.
Ideally, we'd move to having only one reporter, whose lifetime is adjusted
appropriately (from per-run in general to per-context for type checking,
so errors can be buffered -- "silenced" -- during nested type checking calls).

Start the clean up by moving truncation to the REPL,
since it's not relevant for regular reporting. Perversely, we were checking
truncation all the time, even though it's only on during a repl run.
(Truncation is now always turned off in the repl under -verbose.)

Untangle error resetting on symbols from error reporting (reportAdditionalErrors).
This fixes a nice&subtle bug that caused feature warnings to be suppressed under
`-Xfatal-warnings`:

```
def reportCompileErrors() {
  if (!reporter.hasErrors && reporter.hasWarnings && settings.fatalWarnings)
    globalError("No warnings can be incurred under -Xfatal-warnings.")

  if (reporter.hasErrors) { ... }
  else {
    // will erroneously not get here if
    // `reporter.hasWarnings && settings.fatalWarnings`
    // since the `globalError` call above means `reporter.hasErrors`...
    allConditionalWarnings foreach (_.summarize())
    ...
  }
}
```

The second `if`'s condition depends on the `globalError` call in the first `if`...

adriaanm committed:
Move reporting logic into Reporting trait

Move code from Global/SymbolTable to separate Reporting traits to
start carving out an interface in scala.reflect.internal.Reporting,
with internals in scala.tools.nsc. Reporting is mixed into the cake.
It contains a nested class PerRunReporting.

Should do the same for debugging/logging.

The idea is that CompilationUnit and Global forward all reporting
to Reporter. The Reporting trait contains these forwarders, and
PerRunReporting, which accumulates warning state during a run.

In the process, I slightly changed the behavior of `globalError`
in reflect.internal.SymbolTable: it used to abort, weirdly.
I assume that was dummy behavior to avoid introducing an abstract method.
It's immediately overridden in Global, and I couldn't find any other subclasses,
so I don't think the behavior in SymbolTable was ever observed.

Provide necessary hooks for scala.reflect.macros.Parsers#parse.
See scala/reflect/macros/contexts/Parsers.scala's parse method,
which overrides the reporter to detect when parsing goes wrong.
This should be refactored, but that goes beyond the scope of this PR.

Don't pop empty macro context stack.
(Ran into this while reworking -Xfatal-warnings logic.)

Fix -Xfatal-warnings behavior (and check files): it wasn't meant to
influence warning reporting, except for emitting one final error;
if necessary to fail the compile (when warnings but no errors were reported).
Warnings should stay warnings.

This was refactored in fbbbb22946, but we soon seem to have relapsed.
An hour of gitfu did not lead to where it went wrong. Must've been a merge.

som-snytt commented:
review by @som-snytt.  I've been looking for a review opportunity that doesn't involve skolems or some combination of dealiasing and widening.

adriaanm commented:
Great, thank you! I'm still using this PR to check integration. Don't let that stop you.
 The initial refactoring is mostly done % polish.

som-snytt commented:
Someone was complaining or explaining recently about % being remainder instead of modulo.  So do you really mean an equivalence class of polish?  I've been indisposed by employment; punching a clock makes clock arithmetic compelling again.

adriaanm committed:
Uniformly route reporting through reporter.

Sharpen interfaces, reduce footprint of Reporting trait.
Ideally, all reporting should indirect through reporter,
and the `Reporting` trait itself should be restricted to
a single method that retrieves the current `reporter`.

Pull up some more reporting to reflect.internal.
Would like to do more, but need to move partest to the
reflect.internal interface first.
(Its `errorCount` relies on `ERROR.count` in `tools.nsc.Reporter`.)

adriaanm commented:
Well, I think this is mostly ready for reviewing. Fingers crossed for the last two commits.

som-snytt commented:
Great! Note that Jenkins always preempts any attempt to preempt comments.

adriaanm commented:
Re-pre-empting.

adriaanm commented:
Final rebase.

som-snytt commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -227,20 +228,7 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
    */
   def registerTopLevelSym(sym: Symbol) {}
 
-// ------------------ Reporting -------------------------------------
```
IDE needs "refactor commented block to trait".

som-snytt commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -227,20 +228,7 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
-// ------------------ Reporting -------------------------------------
-
-  // not deprecated yet, but a method called "error" imported into
-  // nearly every trait really must go.  For now using globalError.
```
Nobody else dislikes the asymmetry of globalError as an identifier?  After all this time, I don't have anything on the tip of my tongue, but maybe `errorAt(msg, pos = NoPosition)`.  Not sure the regularity of `warnAt` etc is worth it.  But since `Positions` is our dependency, we are always `at`.  The motivation of the comment is that only reporter warns or informs, but lots of funcs like to have an `error` func.

adriaanm replied:
The only solution, I think, is to remove all of these and always indirect through `reporter` (or through `typer.context`).

som-snytt commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -1112,45 +1093,41 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
-        "call site"          -> (site.fullLocationString + " in " + site.enclosingPackage)
-      )
-      ("\n  " + errorMessage + "\n" + info1) :: info2 :: context_s :: Nil mkString "\n\n"
+  def supplementTyperState(errorMessage: String): String = try {
```
Kind of like `freshName`, we need `freshNameForRefactor`, which would generate `supplementErrorMessage$adriaanm$3820$1`.

som-snytt commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -0,0 +1,131 @@
+  class PerRunReporting {
+    // NOTE: scala.reflect.macros.Parsers#parse relies on everything related to reporting going through this def...
+    // TODO: can we rework this to avoid the indirection/fragility?
+    def reporter = Reporting.this.reporter
```
note to self: address the semi-rhetorical question, which is a `[30]` in Knuth notation, i.e., 2 hrs + margin if hulu.com is playing in another window.

som-snytt commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -0,0 +1,131 @@
+        else if (!(warnings contains pos)) warnings += ((pos, msg))
+      def summarize() =
+        if (warnings.nonEmpty && (option.isDefault || settings.fatalWarnings))
+          reporter.warning(NoPosition, "there were %d %s warning(s); re-run with %s for details".format(warnings.size, what, option.name))
```
Confirming that this obviates scrolling right in the diff view on github:

```
reporter.warning(NoPosition, f"there were $size $what warning(s); re-run with ${option.name} for details"
```

adriaanm committed:
Route type checker reporting through context

Continue the work started in SI-8450
(no "implicit numeric widening" warning in silent mode),
which was caused by going straight to the reporter instead of
using the context for type error reporting (which buffers in silent mode).

Ideally, this mistake should not be possible: typer should change
the current reporter to buffer where appropriate.

adriaanm committed:
Rip out reporting indirection from CompilationUnit

Inline the forwarders from CompilationUnit, which should not affect behavior.
Since all forwarders lead to global.reporter, don't first navigate
to a compilation unit, only to then forward back to global.reporter.

The cleanup in the previous commits revealed a ton of confusion
regarding how to report an error.

This was a mechanical search/replace, which has low potential for messing
things up, since the list of available methods are disjoint between
`reporter` and `currentRun.reporting`. The changes involving `typer.context`
were done previously.

Essentially, there are three ways to report:
  - via typer.context, so that reporting can be silenced (buffered)
  - via global.currentRun.reporting, which summarizes (e.g., deprecation)
  - via global.reporter, which is (mostly) stateless and straightforward.

Ideally, these should all just go through `global.currentRun.reporting`,
with the typing context changing that reporter to buffer where necessary.

After the refactor, these are the ways in which we report (outside of typer):
  - reporter.comment
  - reporter.echo
  - reporter.error
  - reporter.warning
  - currentRun.reporting.deprecationWarning
  - currentRun.reporting.incompleteHandled
  - currentRun.reporting.incompleteInputError
  - currentRun.reporting.inlinerWarning
  - currentRun.reporting.uncheckedWarning

Before:
  - c.cunit.error
  - c.enclosingUnit.deprecationWarning
  - context.unit.error
  - context.unit.warning
  - csymCompUnit.warning
  - cunit.error
  - cunit.warning
  - currentClass.cunit.warning
  - currentIClazz.cunit.inlinerWarning
  - currentRun.currentUnit.error
  - currentRun.reporting
  - currentUnit.deprecationWarning
  - currentUnit.error
  - currentUnit.warning
  - getContext.unit.warning
  - getCurrentCUnit.error
  - global.currentUnit.uncheckedWarning
  - global.currentUnit.warning
  - global.reporter
  - icls.cunit.warning
  - item.cunit.warning
  - reporter.comment
  - reporter.echo
  - reporter.error
  - reporter.warning
  - reporting.deprecationWarning
  - reporting.incompleteHandled
  - reporting.incompleteInputError
  - reporting.inlinerWarning
  - reporting.uncheckedWarning
  - typer.context.unit.warning
  - unit.deprecationWarning
  - unit.echo
  - unit.error
  - unit.incompleteHandled
  - unit.incompleteInputError
  - unit.uncheckedWarning
  - unit.warning
  - v1.cunit.warning

All these methods ended up calling a method on `global.reporter`
or on `global.currentRun.reporting` (their interfaces are disjoint).

Also clean up `TypeDiagnostics`: inline nearly-single-use private methods.

adriaanm committed:
Encapsulate deprecation warning message synthesis.

Both refchecks and typer constructed the same message. But different.
Now with more DRYness.

Note that no check files had to be updated (disconcerting)...

adriaanm committed:
Remove deprecationWarning, currentReporting from Reporting

This moves us a bit closer to the goal of having a single
entry point to reporting.

Must modularize Reporting a bit so it can be used in Variances
(need a reference to `currentRun` in `reflect.internal.Reporting`).

som-snytt commented:
@adriaanm sorry I've been overloaded this week. I'd hoped world travel would keep you busy a while longer.  I built it locally and wanted to exercise it, esp from REPL side, which I will do soon. But LGSF.

adriaanm commented:
Thanks, no problem -- I'm in old country for a couple months, so jetlag didn't slow me down this time. There's also no extreme rush.

adriaanm commented:
The push -f strictly rewords/squashes commits, adds comments (based on feedback by @gkossakowski)

adriaanm commented:
All failures were spurious, but there's no way to tell the buildbot about the successful rebuilds. See for yourself by clicking on the unhappy icons next to the commits.

adriaanm commented:
Btw, no need to build locally:

```
$ sbt
> set resolvers += "pr-scala" at "http://private-repo.typesafe.com/typesafe/scala-pr-validation-snapshots/"
> set scalaVersion := "2.11.2-${sha take 7}-SNAPSHOT"
> console
```

Or build your own dist: https://github.com/adriaanm/binfu/blob/master/scafu.sh#L41

adriaanm committed:
Use countElementsAsString for summarized warnings.

adriaanm commented:
Rebased, incorporated the error message grammar improvement, aligned it with our existing counting scheme. Would like to merge this as soon as validation passes.

adriaanm commented:
Aaaand, once^W twice more.

som-snytt commented:
Today, us folks are just sitting around eating hot dogs, drinking beer, watching fireworks, and reviewing code.

adriaanm commented:
Great use of your independence/freedom. Looks like tests are about to go fireworks.

adriaanm commented:
:fireworks:

som-snytt commented:
"Ooooh.  Aaaaaaw!"
